File: day7.py
def parseLine(s):
    parts = s.split(":")
    number_before_colon = int(parts[0].strip())  # Convert the part before the colon to an integer
    numbers_after_colon = list(map(int, parts[1].strip().split()))  # Split and convert the part after the colon to integers
    return number_before_colon, numbers_after_colon

def isValid(lhs, rhs, currVal, currIdx):
    if currIdx == len(rhs):
        return lhs == currVal

    return (isValid(lhs, rhs, currVal * rhs[currIdx], currIdx+1)
            or isValid(lhs, rhs, currVal + rhs[currIdx], currIdx + 1)
            or isValid(lhs, rhs, int(str(currVal) + str(rhs[currIdx])), currIdx + 1))

tot = 0
while line := input():
    lhs, rhs = parseLine(line)
    if isValid(lhs, rhs, rhs[0], 1):
        tot += lhs
print("tot", tot)

File: day7.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <sstream>

using namespace std;

pair<long long, vector<int>> parseLine(string s) {
    vector<int> rhs;

    // Split the string around the colon
    size_t colonPos = s.find(':');
    // Parse the number before the colon
    long long lhs = stoll(s.substr(0, colonPos));

    // Parse the numbers after the colon
    istringstream rhsStream(s.substr(colonPos + 1));
    int num;
    while (rhsStream >> num) {
        rhs.push_back(num);
    }

    return make_pair(lhs, rhs);
}

void printVec(vector<int> v) {
    for (int val : v) cout << val << ' ';
    cout << endl;
}

bool isValid(long long lhs, const vector<int>& rhs, long long currVal, int currIndex) {
    if (currIndex == rhs.size()) {
        return currVal == lhs;
    }

    return isValid(lhs, rhs, currVal * static_cast<long long>(rhs[currIndex]), currIndex + 1)
        || isValid(lhs, rhs, currVal + static_cast<long long>(rhs[currIndex]), currIndex + 1)
        || isValid(lhs, rhs, stoll(to_string(currVal) + to_string(rhs[currIndex])), currIndex + 1);
}

int main() {
    string line;
    long long sum = 0;
    while (getline(cin, line)) {
        if (line == "") break;
        auto data = parseLine(line);
        auto lhs = data.first;
        auto& rhs = data.second;

        if (isValid(lhs, rhs, rhs[0], 1)) {
            sum += lhs;
        }
    }
    cout << "Sum: " << sum << endl;
    return 0;
}
